drivers/volume/portworx/portworx.go

```diff
@@ -1778,6 +1781,287 @@ func (d *portworx) ValidatePureVolumesNoReplicaSets(volumeName string, params ma
+	}
+
+	lsblkParsed, err := parseLsblkOutput(out)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse lsblk output on node %s, Err: %v", n.MgmtIp, err)
+	}
+
+	// Also, raise an error if there is a mismatch in the devices in the two outputs (e.g. if there is a mapper
+	// in dmsetup but not lsblk, something is very wrong and we didn't clean up right)
+	for _, mapper := range dmsetupFoundMappers {
+		lsblkEntry, exists := lsblkParsed[mapper]
+		if !exists {
+			return nil, fmt.Errorf("found mapper %s in dmsetup but not lsblk on node %s, inconsistent disk state (we didn't clean something up right?)", mapper, n.MgmtIp)
+		}
+		// Ensure there are no WWIDs with empty single paths, as that is a malformed device
+		if len(lsblkEntry.SinglePaths) == 0 {
+			return nil, fmt.Errorf("found mapper %s with no single paths on node %s, this should not happen (we didn't clean something up right?)", mapper, n.MgmtIp)
+		}
+	}
+
+	return lsblkParsed, nil
+}
+
+func (d *portworx) getCurrentPureLocalVolumePaths() (map[string]map[string]pureLocalPathEntry, error) {
+	// Iterate through all nodes, and for each node collect local paths
+	currentDevices := make(map[string]map[string]pureLocalPathEntry)
+	nodes := node.GetStorageDriverNodes()
+	for _, n := range nodes {
+		localPathMap, err := d.collectLocalNodeInfo(n)
+		if err != nil {
+			return nil, err
+		}
+		currentDevices[n.MgmtIp] = localPathMap
+	}
+
+	return currentDevices, nil
+}
+
+// InitializePureLocalVolumePaths sets the baseline for how many Pure devices are attached to each node, so that we can compare against it later
+func (d *portworx) InitializePureLocalVolumePaths() error {
+	currentDevices, err := d.getCurrentPureLocalVolumePaths()
+	if err != nil {
+		return err
+	}
+	logrus.Infof("Pure device baseline initialized to: %+v", currentDevices)
+	d.pureDeviceBaseline = currentDevices
+
+	// TODO: we should only include non-FACD drives here. Eventually, we should handle FACD as its own kind of volume in ValidatePureLocalVolumePaths and validate it matches the pxctl cd l output
+
+	return nil
+}
+
+// ValidatePureLocalVolumePaths checks that the given volumes all have the proper local paths present, *and that no other unexpected ones are present*
+func (d *portworx) ValidatePureLocalVolumePaths() error {
+	t := func() error {
+		currentDevices, err := d.getCurrentPureLocalVolumePaths()
+		if err != nil {
+			return err
+		}
+		logrus.Infof("Current pure devices: %+v", currentDevices)
+
+		// TODO: handle FACD drive failovers properly (these will show up as new devices but they're actually fine)
+		// Remove all devices that are in the baseline (complex set math time!). Also warn if any baseline devices are now missing?
+		for node, baselineDevices := range d.pureDeviceBaseline {
+			currentNodeDevices := currentDevices[node]
+			for wwid := range baselineDevices {
+				if _, exists := currentNodeDevices[wwid]; exists {
```

Comment:
Should we compare the single paths here too? We want to make sure the number of singlePaths match before and after.

Reply:
We can but I think we tried this in PSO tests in the pass and we got pretty often false alarms due to temporary path failures - maybe if what we have so far proves to be very consistent we can add it later?